docs/proposals/15-manage-bastions-and-ssh-key-pair-rotation.md

```diff
@@ -0,0 +1,213 @@
+  - Runs on seed (of course)
+- Gardener Controller Manager (`GCM`)
+  - `Bastion` heartbeat controller
+    - Cleans up `Bastion` resource on missing heartbeat.
+    - Is configured with a `maxLifetime` and `timeToLife` for the `Bastion` resource
+- Gardener (RBAC)
+  - The project `admin` role should be extended to allow CRUD operations on the `Bastion` resource. The `gardener.cloud:system:project-member-aggregation` `ClusterRole` needs to be updated accordingly (https://github.com/gardener/gardener/blob/master/charts/gardener/controlplane/charts/application/templates/rbac-user.yaml)
+
+### SSH Flow
+0. Users should only get the RBAC permission to `create` / `update` `Bastion` resources for a namespace, if they should be allowed to `ssh` onto the shoot nodes in this namespace. A project member with `admin` role will have these permissions.
+1. User/`gardenctlv2` creates `Bastion` resource in garden cluster (see [resource example](#resource-example) below)
+    - First, gardenctl would figure out the own public IP of the user's machine. Either by calling an external service (gardenctl (v1) uses https://github.com/gardener/gardenctl/blob/master/pkg/cmd/miscellaneous.go#L226) or by calling a binary that prints the public IP(s) to stdout. The binary should be configurable. The result is set under `spec.ingress[].from[].ipBlock.cidr`
+    - the public `ssh` key of the user is set under `spec.sshPublicKey`. The key that should be used needs to be configured beforehand by the user
+    - The targeted shoot is set under `spec.shootRef`
+2. GAPI Admission Control for the `Bastion` resource in the garden cluster
+    - Mutating Webhook
+        - on creation, sets `metadata.annotations["operations.gardener.cloud/created-by"]` according to the user that created the resource
+        - when `operations.gardener.cloud/operation: keepalive` is set, calculates new `status.expirationTimestamp`
+3. `gardenlet`
+    - Watches `Bastion` resource for own seed under api group `operations.gardener.cloud` in the garden cluster
+    - Creates `Bastion` custom resource under api group `extensions.gardener.cloud/v1alpha1` in the seed cluster
+4. `GCM`:
+    - During reconcile of the `Bastion` resource:
+      - according to `spec.shootRef`, sets the `status.seedName`
+      - according to `spec.shootRef`, sets the `status.providerType`
+5. Gardener extension provider <infra> / Bastion Controller on Seed:
+    - With own `Bastion` Custom Resource Definition in the seed under the api group `extensions.gardener.cloud/v1alpha1`
+    - Watches `Bastion` custom resources that are created by the `gardenlet` in the seed
+    - Controller reads `cloudprovider` credentials from seed-shoot namespace
+    - Deploy infrastructure resources
+        - Bastion VM. User data similar to https://github.com/gardener/gardenctl/blob/1e3e5fa1d5603e2161f45046ba7c6b5b4107369e/pkg/cmd/ssh.go#L160-L171. Writes `spec.sshPublicKey` into `authorized_keys` file.
```

Comment:
Where is this user data coming from? I doubt that we want to duplicate this in every extension provider controller. Should there be another `.spec` field with the user data which is populated by gardenlet, for example (just brain-storming here)?

Reply:
hmm yes, in theory we could then drop `spec.sshPublicKey` on the Bastion resource in the seed and instead have a `spec.userData`, which, as you say, is populated by gardenlet together with `spec.sshPublicKey` from the Bastion resource in the garden cluster. I will have to sleep over it. Or maybe someone else has a different idea

Reply:
Yeah, I think this sounds reasonable. WDYT @xrstf?

Reply:
So far it looked like the userdata was minimal (https://github.com/gardener/gardenctl/blob/1e3e5fa1d5603e2161f45046ba7c6b5b4107369e/pkg/cmd/ssh.go#L160-L171), so I don't think it hurts too much to have every provider ship its own userdata string. At least this is somewhat manageable.

If we opened the door to allow arbitrary userdata on Bastion hosts, won't this increase the attack surface and could lead to other sorts of issues?

Reply:
Hm, not sure, this would only be possible in the seed cluster. The resource in the garden cluster would not allow to provide arbitrary user data. Even if it's minimal today, maybe it grows tomorrow. Or, if there is a change, we would have to touch `n` providers instead of only once (if it was in the gardenlet).

Reply:
replaced `spec.sshPublicKey` with `spec.userData` on Bastion resource in seed https://github.com/gardener/gardener/pull/3802/commits/5c2004a1d49946230cf42460876f22fe2832e232

This also helps in case we choose to implement the suggestion from @mvladev https://github.com/gardener/gardener/pull/3802#pullrequestreview-637556223 , where instead of the users public ssh key the ssh-ca public key would be added as `TrustedUserCAKeys` to the bastion. The extension provider would not care what is inside the user data, as this will be managed / populated by the gardenlet.